### DISCUSSION TITLE
mypy - exclude option

### DISCUSSION THREAD

**Discussion by mayk0gan** [akaihola/darker#346] [Q&A] [answered]
Hi, quoted from the README file:

> Darker does honor exclusion options in Black configuration files when recursing directories, but the exclusions are only applied to Black reformatting. Isort and linters are still run on excluded files. Also, individual files explicitly listed on the command line are still reformatted even if they match exclusion patterns.

"Isort and linters are still run on excluded files" - is there a particular reason for that?
We use mypy's `exclude` option - which being ignored here

**Comment by akaihola** [comment, marked as answer]
Thanks for the great question @mayk0gan!

The exclusion options passed to Black come from Black configuration files, not Darker's command line options or `[tool.darker]` in `pyproject.toml`.

It just happened to be that in #146 @roniemartinez only needed this to skip Black reformatting, and we made the design decision to only do that and not touch isort and linter behavior.

If we want to be able to exclude files "globally" (for Black, isort and all linters), I think we need to come up with a Darker option which can be passed on the command line or specified in `pyproject.toml`.

The reason `mypy` doesn't honor the `exclude` option when run with Darker is that Darker calls `mypy` for each file individually. For example if I have `exclude = ['src/']` in the `[mypy]` section of `mypy.ini`:
- running `mypy .` will skip all files in `src/`
- running `mypy src` will skip all files
- running `mypy src/darker` will skip all files
- running `mypy src/darker/tests` will skip all files
- but running `mypy src/darker/*.py` won't skip the matching files
- and running `mypy src/darker/config.py` won't skip the only listed file, and this is how Darker calls `mypy`

There's actually a plan to change how `mypy` is run by Darker, for different reasons. See #328. I think we'll change to run `mypy` by passing it all the same files/directories which Darker receives on its command line. So running e.g. `darker -L mypy src/` would then cause `mypy` to actually honor the `exclude` option. But for the IDE use case this wouldn't probably help, since IDEs tend to call Darker with one file at a time.

Does this make sense?

**Reply by roniemartinez** [reply]
To add here, I only requested for Black config support in #146 to keep things DRY and at the time the ticket was opened we are not using Darker completely (not all devs in my team use it). I didn't want to overexert the effort done by @akaihola and the maintainers.

isort and mypy config files can be supported with the idea of keeping things DRY. I believe we can just open a new feature request if people needs it.

**Reply by mayk0gan** [reply]
Thanks for your great answer @akaihola!
Waiting for https://github.com/akaihola/darker/discussions/328 :)

cc @saare
